### PR TITLE
Alex Thurston - UI Design Patterns Challenge

### PR DESCRIPTION
**Pull request by terranisaur**

### CODE REVIEW

File: MVVM Module/Source/Feed UI/Controllers/FeedRefreshViewController.swift [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	var viewModel: FeedViewModel? {
+		didSet { bind() }
+	}
+
+	@IBAction func refresh() {
+		viewModel?.loadFeed()
+	}
+
+	func bind() {
+		viewModel?.onLoadingStateChange = { [weak self] loadingState in
```
Comment (thiagopenna) [inline MVVM Module/Source/Feed UI/Controllers/FeedRefreshViewController.swift:25]: Following my previous comment, we can implement a `viewModel?.onErrorStateChange`, and update the `self?.errorView?` accordingly, instead of combining both behaviors inside the `loading` logic.

File: MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift [hunk 2]
```diff
@@ -77,6 +77,26 @@ final class FeedUIIntegrationTests: XCTestCase {
 		assertThat(sut, isRendering: [image0])
 	}
 
+	func test_loadFeedCompletion_doesNotShowLoadFeedError() {
```
Comment (thiagopenna) [inline MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift:80]: These tests are not asserting the behavior that starting a manual refresh dismisses the error message, so there's code that's not covered by them. Take a look at `FeedUIIntegrationTests` inside the `MVC` module and how `test_loadFeedCompletion_rendersErrorMessageOnErrorUntilNextReload` is implemented to check for that.

File: MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift [hunk 3]
```diff
@@ -77,6 +77,26 @@ final class FeedUIIntegrationTests: XCTestCase {
...
+
+		XCTAssertFalse(sut.isShowingLoadingError)
+	}
+
+	func test_loadFeedError_tapToDismissLoadingError() {
+		let (sut, loader) = makeSUT()
+
+		sut.loadViewIfNeeded()
+		loader.completeFeedLoadingWithError()
+		XCTAssertTrue(sut.isShowingLoadingError)
```
Comment (thiagopenna) [inline MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift:94]: To improve this test it's possible to check the error message being displayed, to ensure the `sut` is showing the correct error message.
```suggestion
		XCTAssertEqual(sut.errorMessage, localized("FEED_VIEW_CONNECTION_ERROR"))
```

File: MVVM Module/Source/Feed UI/Controllers/FeedRefreshViewController.swift [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+//  Created by Alex Thurston on 6/16/21.
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import UIKit
+import FeedFeature
+
+final class FeedRefreshViewController: NSObject {
+	@IBOutlet var view: UIRefreshControl?
+	@IBOutlet var errorView: ErrorView?
```
Comment (thiagopenna) [inline MVVM Module/Source/Feed UI/Controllers/FeedRefreshViewController.swift:14]: The setter for this property can be private to prevent clients from changing the instance:

```suggestion
	@IBOutlet private(set) var errorView: ErrorView?
```

File: MVVM Module/Source/Feed UI/Composers/FeedUIComposer.swift [hunk 5]
```diff
@@ -14,6 +14,7 @@ public final class FeedUIComposer {
 
 		let feedController = FeedViewController.makeWith(
 			viewModel: feedViewModel)
+		feedController.refreshController?.viewModel = feedViewModel
```
Comment (thiagopenna) [inline MVVM Module/Source/Feed UI/Composers/FeedUIComposer.swift:17]: If there is a need to change the creation of a `FeedViewController`, it should be altered inside the `makeWith(viewModel:)` method (especially considering it already receives the viewModel as a parameter). 

Although, in this case there's no need to change `refresh` logic, so following from my previous comments please revert this change regarding `FeedRefreshViewController`.
```suggestion
```

File: MVVM Module/Tests/Feed UI/FeedUIIntegrationTests.swift [hunk 6]
```diff
@@ -77,6 +77,26 @@ final class FeedUIIntegrationTests: XCTestCase {
 		assertThat(sut, isRendering: [image0])
 	}
 
+	func test_loadFeedCompletion_doesNotShowLoadFeedError() {
```
Comment (thiagopenna) [inline MVVM Module/Tests/Feed UI/FeedUIIntegrationTests.swift:80]: The same comments from the `MVP` module apply in these tests, see my other comment regarding untested behavior and how to improve coverage.

**Review by thiagopenna — COMMENTED** [review]
Hi @terranisaur, you almost passed the challenge!

This is Penna, I am an instructor at Essential Developer.

Please, check my review comments and let me know once you're done so I can review it again. 

Thank you 🙂

**Comment by terranisaur** [comment]
I pushed my updates. Thanks for the feedback.

File: MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 7]
```diff
@@ -19,19 +21,28 @@ public final class FeedViewController: UITableViewController, UITableViewDataSou
 		refresh()
 	}
 
-	@IBAction private func refresh() {
+	@IBAction func refresh() {
```
Comment (thiagopenna) [inline MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift:24]: There's no need to call `refresh` outside of `FeedViewController`, so this function should be marked as private.
```suggestion
	@IBAction private func refresh() {
```

**Review by thiagopenna — COMMENTED** [review]
Hi @terranisaur,

Congratulations, you've passed the challenge 🥇

- All tests are passing ✅
- Your code is clean and handles edge cases ✅
- The UI and animations look and behave as expected ✅
- Light and Dark mode support ✅
- Added an extra language ✅

I left just one review comment. Feel free to push changes.

Great job!

**Comment by terranisaur** [comment]
I have a question. Have I completed the course after the main module, or after the [Bonus Lives] section? I think I remember there was a certificate that is awarded upon completing the course, do I need to do anything special to get that?

**Comment by caiozullo** [comment]
Hi @terranisaur, you need to complete all challenges and all lectures, including the live sessions, to get the certificate of completion.

**Comment by terranisaur** [comment]
Thanks!
